### PR TITLE
Contrib/customize embeddings

### PR DESCRIPTION
**Pull request by skrawcz**
This examples uses an [openai cookbook](https://github.com/openai/openai-cookbook/blob/main/examples/Customizing_embeddings.ipynb)
as inspiration for a module to customize embeddings.

One can easily pull this module down, create a dataset that fits the expected schema,
and then customize those embeddings easily.

## For new dataflows:
Do you have the following?
- [x] Added a directory mapping to my github user name in the contrib/hamilton/contrib/user directory.
  - [x] If my author names contains hyphens I have replaced them with underscores.
  - [x] If my author name starts with a number, I have prefixed it with an underscore.
  - [x] If your author name is a python reserved keyword. Reach out to the maintainers for help.
  - [x] Added an author.md file under my username directory and is filled out.
  - [x] Added an __init__.py file under my username directory.
- [ ] Added a new folder for my dataflow under my username directory.
  - [x] Added a README.md file under my dataflow directory that follows the standard headings and is filled out.
  - [x] Added a __init__.py file under my dataflow directory that contains the Hamilton code.
  - [x] Added a requirements.txt under my dataflow directory that contains the required packages outside of Hamilton.
  - [x] Added tags.json under my dataflow directory to curate my dataflow.
  - [x] Added valid_configs.jsonl under my dataflow directory to specify the valid configurations.
  - [x] Added a dag.png that shows one possible configuration of my dataflow.
- [x] I hearby acknowledge that to the best of my ability, that the code I have contributed contains correct attribution
and notices as appropriate.

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Dataflow documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: contrib/hamilton/contrib/user/skrawcz/customize_embeddings/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+`processed_local_dataset` as an input. The dataframe should have columns "text_1", "text_2", and "label". The label should be +1 if the
+text pairs are similar and -1 if the text pairs are dissimilar.
+
+In general to use this module you'll need to do the following:
+
+1. Create a dataset of text pairs, where each pair is either similar or dissimilar. Mechanically that means you'll need
+to modify/override `processed_local_dataset()` so that you output a dataset of [text_1, text_2, label] where label is +1 if
+the pairs of text are similar and -1 if the pairs of text are dissimilar. See the docstring for `processed_local_dataset()`.
+2. Modify the "hyperparameters" for a couple of the functions, e.g. `optimize_matrix()`, to suit your needs.
+3. Modify how embeddings are generated, e.g. `_get_embedding()`, to suit your needs.
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/customize_embeddings/README.md:38]: no need to over engineer/scope creep. They can modify the code instead.

File: contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+    return "data/snli_embedding_cache.pkl"  # embeddings will be saved/loaded here
+
+
+def default_embedding_engine() -> str:
+    """Default embedding engine. Use `overrides={"default_embedding_engine": "my_engine"}` to override."""
+    return "babbage-similarity"  # text-embedding-ada-002 is recommended
+
+
+@config.when(source="snli")
+def snli_dataset(download_path: str = "data") -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py:94]: Custom loader for a one off thing to run an example? Maybe for someone wanting an example of an adapter. But from an example perspective I think that's too much for the value of it.

File: contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+    :param dataset_value: train or test.
+    :param embedded_data_set: the embedded data set to compute accuracy on given the dataset_value..
+    :return: tuple of accuracy and standard error
+    """
+    data = embedded_data_set[embedded_data_set["dataset"] == dataset_value]
+    a, se = _accuracy_and_se(data["cosine_similarity"], data["label"])
+    return a, se
+
+
+def _embedding_multiplied_by_matrix(embedding: List[float], matrix: torch.tensor) -> np.array:
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py:457]: we don't have that capability yet.

File: contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+    :param dropout_fraction: From the cookbook - dropout helped by a couple percentage points (definitely not necessary).
+    :param print_progress:
+    :param save_results:
+    :return: dataframe cataloging the optimization results.
+    """
+    run_id = random.randint(0, 2**31 - 1)  # (range is arbitrary)
+    # convert from dataframe to torch tensors
+    # e is for embedding, s for similarity label
+
+    def tensors_from_dataframe(
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py:517]: from the cookbook.

File: contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+
+optimization_parameterization = {
+    "matrix_b10_l10": {"batch_size": value(10), "learning_rate": value(10.0)},
+    "matrix_b100_l100": {"batch_size": value(100), "learning_rate": value(100.0)},
+    "matrix_b1000_l1000": {"batch_size": value(1000), "learning_rate": value(1000.0)},
+}
+
+
+@parameterize(**optimization_parameterization)
+def optimize_matrix(
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py:491]: but this is a `@parameterize` function. So 🤷  could go both ways.

File: contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py [hunk 6]
```diff
@@ -0,0 +1,733 @@
...
+        "text_2": pa.Column(str, nullable=False),
+        "label": pa.Column(int, nullable=False, checks=[pa.Check.isin([-1, 1])]),
+    },
+    strict=True,
+)
+
+
+@config.when(source="snli")
+@check_output(schema=processed_dataset_schema, importance="fail")
+def processed_local_dataset__snli(
```
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply contrib/hamilton/contrib/user/skrawcz/customize_embeddings/__init__.py:133]: If you build the right data loader I think so, but yeah, this is fine

**Pull request merged**